installinator/src/write.rs

```diff
@@ -171,3 +187,140 @@ async fn write_artifact_impl(
 
     Ok(())
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_helpers::{dummy_artifact_hash_id, with_test_runtime};
+
+    use anyhow::Result;
+    use bytes::{Buf, Bytes};
+    use camino::Utf8Path;
+    use futures::StreamExt;
+    use omicron_test_utils::dev::test_setup_log;
+    use proptest::prelude::*;
+    use tempfile::tempdir;
+    use test_strategy::proptest;
+    use tokio::io::AsyncReadExt;
+    use tokio_stream::wrappers::ReceiverStream;
+
+    #[proptest(ProptestConfig { cases: 32, ..ProptestConfig::default() })]
+    fn proptest_write_artifact(
+        #[strategy(prop::collection::vec(prop::collection::vec(any::<u8>(), 0..8192), 0..16))]
+        data: Vec<Vec<u8>>,
+    ) {
+        with_test_runtime(move || async move {
+            proptest_write_artifact_impl(data).await.expect("test failed");
+        })
+    }
+
+    async fn proptest_write_artifact_impl(data: Vec<Vec<u8>>) -> Result<()> {
+        let logctx = test_setup_log("test_write_artifact");
+        let tempdir = tempdir()?;
+        let tempdir_path: &Utf8Path = tempdir.path().try_into()?;
+        let temp_destination = tempdir_path.join("test.bin");
+
+        let artifact_id = dummy_artifact_hash_id();
+        let mut artifact: BufList = data.into_iter().map(Bytes::from).collect();
+
+        let (event_sender, event_receiver) = mpsc::channel(512);
+
+        let receiver_handle = tokio::spawn(async move {
+            ReceiverStream::new(event_receiver).collect::<Vec<_>>().await
+        });
+
+        write_artifact(
+            &logctx.log,
+            &artifact_id,
+            artifact.clone(),
+            &temp_destination,
+            &event_sender,
+        )
+        .await;
+
+        std::mem::drop(event_sender);
+
+        let events = receiver_handle.await?;
+
+        let mut seen_completion = false;
+        let mut last_written_bytes = 0;
+
+        for event in events {
+            match event {
+                ReportEvent::Progress(ProgressEventKind::WriteProgress {
+                    attempt,
+                    kind,
+                    destination,
+                    written_bytes,
+                    total_bytes,
+                    ..
+                }) => {
+                    assert!(
+                        !seen_completion,
+                        "no more progress events after completion"
+                    );
+                    assert_eq!(attempt, 1);
```

Comment:
maybe use prop_assert_eq! so it only displays on shrunken version?

Reply:
Sadly if you do that you get:

```
error[E0308]: mismatched types
   --> installinator/src/write.rs:431:9
    |
431 | /         prop_assert_eq!(
432 | |             read_num_bytes,
433 | |             artifact.num_bytes(),
434 | |             "read num_bytes matches"
435 | |         );
    | |         ^
    | |         |
    | |_________expected struct `anyhow::Error`, found enum `TestCaseError`
    |           arguments to this enum variant are incorrect
```